### PR TITLE
Plugin Request: new plugin for Abema.tv

### PR DESCRIPTION
**Pull request by jackyzy823**
## Plugin Request

- [x] I have read the contribution and plugin request guidelines. 

### Description
Abema.tv  is a japanese live (news,drama,anime , etc) and vod stream site.

### Example stream URLs
1. Live stream:  https://abema.tv/now-on-air/abema-news
2. VOD stream: https://abema.tv/video/episode/90-1053_s99_p12
3. VOD stream: https://abema.tv/channels/everybody-anime/slots/FJcUsdYjTk1rAb

### Comments, screenshots, etc.
1.Because of `ext-x-discontinuity` issue, players may not handle live stream discontinuity properly.
2.Streams are geo-restricted to Japan , but could be bypassed via `--http-header 'X-Forwarded-For=A_JAPAN_IP' `   (TESTONLY)

### CODE REVIEW

File: src/streamlink/plugins/abematv.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+
+class AbemaTV(Plugin):
+    '''
+    Abema.tv https://abema.tv/
+    Note: Streams are geo-restricted to Japan
+
+    '''
+
+    _url_re = re.compile(r"https://abema\.tv/now-on-air/(?P<onair>.+)|https://abema\.tv/video/episode/(?P<episode>.+)|https://abema\.tv/channels/.+?/slots/(?P<slot>.+)")
```
Comment (beardypig) [inline src/streamlink/plugins/abematv.py]: This could be simplified a bit I think.

**Comment by beardypig** [comment]
Looks good, there are a couple of strings that could be made in to class attributes to make it slightly more readable. Also there is a random `pass` in there :)

File: src/streamlink/plugins/abematv.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+            onair = matchResult.group("onair")
+            channels = self.session.http.get("https://api.abema.io/v1/channels").json()["channels"]
+            for channel in channels:
+                if onair == channel["id"]:
+                    break
+            else:
+                raise NoStreamsError 
+            playlisturl = channel["playback"]["hls"]
+        elif matchResult.group("episode"):
+            playlisturl = "https://vod-abematv.akamaized.net/program/%s/playlist.m3u8"%(matchResult.group("episode"))
```
[the remark this answers is not in the document]
Comment (beardypig) [reply src/streamlink/plugins/abematv.py]: Or even better, use `string.format` :)

File: src/streamlink/plugins/abematv.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+
+
+class AbemaTV(Plugin):
+    '''
+    Abema.tv https://abema.tv/
+    Note: Streams are geo-restricted to Japan
+
+    '''
+    _url_re = re.compile(r"https://abema\.tv/(now-on-air/(?P<onair>.+)|video/episode/(?P<episode>.+)|channels/.+?/slots/(?P<slot>.+))")
```
[the remark this answers is not in the document]
Comment (jackyzy823) [reply src/streamlink/plugins/abematv.py]: I think URLs with params could be valid although we just drop it and i add test cases for URLs with params.

**Comment by jackyzy823** [comment]
@beardypig  @back-to Thanks for your reviews.

File: src/streamlink/plugins/abematv.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jackyzy823) [reply src/streamlink/plugins/abematv.py]: In line 121 `url, params = parse_url_params(self.url)`  params are dropped.  

Dropping params in `can_handle_url` too would be better ?
```
    @classmethod
    def can_handle_url(cls, url):
        url, params = parse_url_params(url)
        return cls._url_re.match(url) is not None
```

[the remark this answers is not in the document]
Comment (back-to) [reply src/streamlink/plugins/abematv.py]: it is not

```py
>>> from streamlink.plugin.plugin import parse_url_params
>>> x = "https://abema.tv/video/episode/90-1053_s99_p14?next=true"
>>> url, params = parse_url_params(x)
>>> url
'https://abema.tv/video/episode/90-1053_s99_p14?next=true'
>>> 
```

`parse_url_params` is for something else

```py
>>> x = "https://abema.tv/video/episode/90-1053_s99_p14 next=true"
>>> url, params = parse_url_params(x)
>>> url
'https://abema.tv/video/episode/90-1053_s99_p14'
>>> params
{'next': 'true'}
>>> 
```

[the remark this answers is not in the document]
Comment (jackyzy823) [reply src/streamlink/plugins/abematv.py]: I see . A misuse of this function.
Is this kind of regex acceptable?
```
    _url_re = re.compile(r"""https://abema\.tv/(
        now-on-air/(?P<onair>[^\?]+).*
        |
        video/episode/(?P<episode>[^\?]+).*
        |
        channels/.+?/slots/(?P<slot>[^\?]+).*
        )""", re.VERBOSE)
```

[the remark this answers is not in the document]
Comment (back-to) [reply src/streamlink/plugins/abematv.py]: @jackyzy823 

you don't need the `.*` part

```py
>>> import re
>>> url = "https://abema.tv/video/episode/90-1053_s99_p14?next=true"
>>> _url_re = re.compile(r"""https://abema\.tv/(
...     now-on-air/(?P<onair>[^\?]+)
...     |
...     video/episode/(?P<episode>[^\?]+)
...     |
...     channels/.+?/slots/(?P<slot>[^\?]+)
...     )""", re.VERBOSE)
>>> m = _url_re.match(url)
>>> m
<_sre.SRE_Match object; span=(0, 46), match='https://abema.tv/video/episode/90-1053_s99_p14'>
>>> m.group("episode")
'90-1053_s99_p14'
```

[the remark this answers is not in the document]
Comment (jackyzy823) [reply src/streamlink/plugins/abematv.py]: There are URLs like `https://abema.tv/video/episode/90-1053_s99_p14?utm_source=social&utm_medium=social&utm_campaign=slot_share` in the site. so i think `.*`part is necessary.

[the remark this answers is not in the document]
Comment (back-to) [reply src/streamlink/plugins/abematv.py]: why? it still matches the regex

```py
>>> url2 = "https://abema.tv/video/episode/90-1053_s99_p14?utm_source=social&utm_medium=social&utm_campaign=slot_share"
>>> m = _url_re.match(url2)
>>> m.group("episode")
'90-1053_s99_p14'
```

[the remark this answers is not in the document]
Comment (jackyzy823) [reply src/streamlink/plugins/abematv.py]: Oh.Sorry for misunderstanding regex grammer .You are right.I will change it.

[the remark this answers is not in the document]
Comment (beardypig) [reply src/streamlink/plugins/abematv.py]: `re.match` finds the first match, so unless you have `$` in the the expression it will match :)

**Comment by back-to** [comment]
for videos which require a login,
the plugin will **fail very late** and **it will look like a streamlink issue** rather than that a login is required

it would be better to get an earlier error warning,
but not sure if this is posible for every video.

---

the api could be used for `episode` as they have a free label

`https://abema.tv/video/episode/90-1053_s99_p12`
`https://api.abema.io/v1/video/programs/90-1053_s99_p14`

```json
{
    "label": {
        "free": true
    },
    "freeEndAt": 1563202800,
    "playback": {
        "hls": "https://vod-abematv.akamaized.net/program/90-1053_s99_p14/playlist.m3u8"
    }
}

```

but `slots` does not have a label there which is unfortunate.

`https://abema.tv/channels/everybody-anime/slots/FJcUsdYjTk1rAb`
`https://api.abema.io/v1/media/slots/FJcUsdYjTk1rAb`

**Comment by jackyzy823** [comment]
For slots `https://api.abema.io/v1/media/slots/FJcUsdYjTk1rAb`   ,  `timeshiftFree` is the label
```json
{
  "slot": {
    "flags": {
      "share": true,
      "timeshift": true,
      "sharingPolicy": 1,
      "timeshiftFree": true,
      "chasePlay": true
    }
  }
}
```

**Comment by jackyzy823** [comment]
CI Failed because  required library `iso3166` update to version `0.9` serveral  hours ago. and it changes "United States" to "United States of America" .

**Comment by gravyboat** [comment]
@jackyzy823 I just merged https://github.com/streamlink/streamlink/pull/1950 to master which should address this (thanks again @back-to). If you rebase tests should pass.

**Comment by back-to** [comment]
could you use git rebase instead of git merge

```
git reset --hard 3f9849b9465e4096d9762eb81eb1a9ac3d8dedda
```
> HEAD is now at 3f9849b plugin.abematv: fail fast if video requiring a login and check json schema

```
git rebase master
```

```
git push --force ...
```

**Comment by back-to** [comment]
your `master` was probably to old,

example how it should be
https://github.com/back-to/streamlink/tree/jackyzy823-plugin.abematv
https://github.com/back-to/streamlink/commits/jackyzy823-plugin.abematv

**Comment by jackyzy823** [comment]
Feel so sorry to have you confused. @back-to

**Comment by jackyzy823** [comment]
Is this mergable or any other things i should change?   :)

**Comment by gravyboat** [comment]
We should be good but I want to wait to confirm with @back-to before doing the merge.

**Comment by DonoMit** [comment]
@gravyboat Off topic sorry, but how can i contact you privately about a project i have a interest in?

**Comment by gravyboat** [comment]
@DonoMit If this is related to Streamlink please open an issue on the repo, if it's about any sort of plugin/website/etc. that you want to keep private please don't contact me. Otherwise there's a consulting page on my website (link on my profile page) where you can contact me. Let's not discuss this any further because as you noted it's off topic.

**Comment by gravyboat** [comment]
Thanks @jackyzy823, and thanks everyone else for the reviews.

**Pull request merged**